neighsyncd/neighsync.cpp

```diff
@@ -9,14 +9,273 @@
 #include "ipaddress.h"
 #include "netmsg.h"
 #include "linkcache.h"
+#include "select.h"
 
 #include "neighsync.h"
+#include "warm_restart.h"
 
 using namespace std;
 using namespace swss;
 
-NeighSync::NeighSync(DBConnector *db) :
-    m_neighTable(db, APP_NEIGH_TABLE_NAME)
+const AppRestartAssist::cache_state_map AppRestartAssist::cacheStateMap =
+{
+    {STALE,     "STALE"},
+    {SAME,      "SAME"},
+    {NEW,       "NEW"},
+    {DELETE,    "DELETE"},
+    {UNKNOWN,   "UNKNOWN"}
+};
+
+AppRestartAssist::AppRestartAssist(RedisPipeline *pipelineAppDB,
+    const std::string &app_name, const std::string &docker_name,
+    ProducerStateTable *ps_table, const uint32_t defaultWarmStartTimerValue):
+    m_appTable(pipelineAppDB, APP_NEIGH_TABLE_NAME, false),
+    m_appName(app_name),
+    m_dockerName(docker_name),
+    m_psTable(ps_table),
+    warmStartTimer(timespec{0, 0})
+{
+    WarmStart::checkWarmStart(m_appName, m_dockerName);
+
+    m_appTableName = m_appTable.getTableName();
+
+    // set the default timer value
+    if (defaultWarmStartTimerValue != 0)
+    {
+        m_reconcileTimer = defaultWarmStartTimerValue;
+    }
+
+    if (!WarmStart::isWarmStart())
+    {
+        m_warmStartInProgress = false;
+    }
+    else
+    {
+        m_warmStartInProgress = true;
+        uint32_t temp_value = WarmStart::getWarmStartTimer(m_appName, m_dockerName);
+        if (temp_value != 0)
+        {
+            m_reconcileTimer = temp_value;
+        }
+
+        warmStartTimer.setInterval(timespec{m_reconcileTimer, 0});
+
+        // Clear the producerstate table to make sure no pending data for the AppTable
+        m_psTable->clear();
+
+        WarmStart::setWarmStartState(m_appName, WarmStart::INIT);
+    }
+}
+
+AppRestartAssist::~AppRestartAssist()
+{
+}
+
+/* join the field-value strings for straight printing */
+string AppRestartAssist::joinVectorString(const vector<FieldValueTuple> &fv)
```

Comment:
>joinVectorString [](start = 25, length = 16)

You could write a general join function and move it to swss-common. There is one more ", " at the end.

Reply:
Yes, it was discussed, let's move this to later PR. thx